File: android2_lesson06_2/src/java/java/com/example/ama/android2_lesson06_2/ui/list/DeviceListFragment.java
package com.example.ama.android2_lesson06_2.ui.list;

import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v7.app.AppCompatActivity;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.ListView;
import android.widget.ProgressBar;
import android.widget.Switch;

import com.example.ama.android2_lesson06_2.R;
import com.example.ama.android2_lesson06_2.ui.MainActivity;
import com.example.ama.android2_lesson06_2.ui.Navigator;
import com.example.ama.android2_lesson06_2.ui.list.adapter.DeviceListAdapter;

public class DeviceListFragment extends Fragment {
    private final static int ENABLE_REQUEST_CODE = 1;

    private Switch bluetoothSwitch;
    private Button buttonFind;
    private ProgressBar progressBar;

    private BluetoothAdapter bluetoothAdapter;
    private DeviceListAdapter adapter;
    private BroadcastReceiver receiver;

    public static DeviceListFragment newInstance() {
        return new DeviceListFragment();
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_device_list, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        initUI(view);
        initBluetooth();
        initReceiver();
        super.onViewCreated(view, savedInstanceState);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, @Nullable Intent data) {
        if (requestCode == ENABLE_REQUEST_CODE)
            if (resultCode == AppCompatActivity.RESULT_CANCELED) {
                bluetoothSwitch.setChecked(false);
                buttonFind.setEnabled(false);
            } else if (resultCode == AppCompatActivity.RESULT_OK) {
                buttonFind.setEnabled(true);
            }
    }

    private void initUI(View view) {
        progressBar = view.findViewById(R.id.pb_progress);
        buttonFind = view.findViewById(R.id.btn_find_devices);
        buttonFind.setOnClickListener(v -> findDevices());
        bluetoothSwitch = view.findViewById(R.id.s_bluetooth);
        bluetoothSwitch.setOnCheckedChangeListener((compoundButton, checked) -> {
            if (checked) enableBluetooth();
            else disableBluetooth();
        });
        if (getActivity() != null) {
            adapter = new DeviceListAdapter(getActivity());
            ListView listView = view.findViewById(R.id.lv_devices);
            listView.setAdapter(adapter);
            listView.setOnItemClickListener((adapterView, view1, i, l) ->
                    Navigator.plaseDetailsListFragment(
                            (MainActivity) getActivity(),
                            (BluetoothDevice) adapterView.getItemAtPosition(i)));
        }
    }

    private void initBluetooth() {
        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        if (bluetoothAdapter == null) {
            bluetoothSwitch.setEnabled(false);
            buttonFind.setEnabled(false);
        } else if (bluetoothAdapter.isEnabled()) {
            bluetoothSwitch.setChecked(true);
            buttonFind.setEnabled(true);
        }
    }

    private void enableBluetooth() {
        if (!bluetoothAdapter.isEnabled()) {
            Intent intent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
            startActivityForResult(intent, ENABLE_REQUEST_CODE);
        }
    }

    private void disableBluetooth() {
        if (bluetoothAdapter.isEnabled()) {
            bluetoothAdapter.disable();
            buttonFind.setEnabled(false);
        }
    }

    private void findDevices() {
        bluetoothAdapter.startDiscovery();
        adapter.clear();
    }

    private void initReceiver() {
        receiver = new BroadcastReceiver() {
            @Override
            public void onReceive(Context context, Intent intent) {
                if (intent.getAction() != null) {
                    switch (intent.getAction()) {
                        case BluetoothDevice.ACTION_FOUND:
                            BluetoothDevice device = intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE);
                            adapter.addDevice(device);
                            break;
                        case BluetoothAdapter.ACTION_DISCOVERY_STARTED:
                            progressBar.setVisibility(View.VISIBLE);
                            break;
                        case BluetoothAdapter.ACTION_DISCOVERY_FINISHED:
                            progressBar.setVisibility(View.GONE);
                            break;
                    }
                }
            }
        };
        IntentFilter filter = new IntentFilter();
        filter.addAction(BluetoothDevice.ACTION_FOUND);
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_STARTED);
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED);
        if (getActivity() != null) {
            getActivity().registerReceiver(receiver, filter);
        }
    }

    @Override
    public void onDestroy() {
        if (getActivity() != null) {
            getActivity().unregisterReceiver(receiver);
        }
        super.onDestroy();
    }
}

File: android2_lesson06_2/src/kotlin/java/com/example/ama/android2_lesson06_2/ui/list/DeviceListFragment.kt
package com.example.ama.android2_lesson06_2.ui.list

import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Bundle
import android.support.v4.app.Fragment
import android.support.v7.app.AppCompatActivity
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.example.ama.android2_lesson06_2.R
import com.example.ama.android2_lesson06_2.ui.MainActivity
import com.example.ama.android2_lesson06_2.ui.Navigator
import com.example.ama.android2_lesson06_2.ui.list.adapter.DeviceListAdapter
import kotlinx.android.synthetic.main.fragment_device_list.*

class DeviceListFragment : Fragment() {
    private var bluetoothAdapter: BluetoothAdapter? = null
    private var adapter: DeviceListAdapter? = null

    private val receiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            if (intent?.action != null) {
                when (intent.action) {
                    BluetoothDevice.ACTION_FOUND -> {
                        val device: BluetoothDevice =
                                intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
                        adapter?.addDevice(device)
                    }
                    BluetoothAdapter.ACTION_DISCOVERY_STARTED -> pb_progress.visibility = View.VISIBLE
                    BluetoothAdapter.ACTION_DISCOVERY_FINISHED -> pb_progress.visibility = View.GONE
                }
            }
        }
    }

    companion object {
        const val ENABLE_REQUEST_CODE = 1

        fun newInstance(): DeviceListFragment {
            return DeviceListFragment()
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        retainInstance = true
        super.onCreate(savedInstanceState)
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.fragment_device_list, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        initList()
        setListeners()
        initBluetooth()
        initReceiver()
        if (bluetoothAdapter != null && bluetoothAdapter!!.isDiscovering) {
            pb_progress.visibility = View.VISIBLE
        }
        super.onViewCreated(view, savedInstanceState)
    }

    override fun onDestroyView() {
        activity?.unregisterReceiver(receiver)
        super.onDestroyView()
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == ENABLE_REQUEST_CODE)
            if (resultCode == AppCompatActivity.RESULT_CANCELED) {
                s_bluetooth.isChecked = false
                btn_find_devices.isEnabled = false
            } else if (resultCode == AppCompatActivity.RESULT_OK) {
                btn_find_devices.isEnabled = true
            }
    }

    private fun initList() {
        if (activity != null) {
            if (adapter == null) {
                adapter = DeviceListAdapter(activity!!)
            }
            lv_devices.adapter = adapter
            lv_devices.setOnItemClickListener { adapterView, view, i, l ->
                if (activity != null) {
                    Navigator.placeDetailsListFragment(activity as MainActivity,
                            adapterView.getItemAtPosition(i) as BluetoothDevice)
                }
            }
        }
    }

    private fun setListeners() {
        btn_find_devices.setOnClickListener { v -> findDevices() }
        s_bluetooth.setOnCheckedChangeListener { _, checked ->
            if (checked) enableBluetooth() else disableBluetooth()
        }
    }

    private fun initBluetooth() {
        bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        if (bluetoothAdapter == null) {
            s_bluetooth.isEnabled = false
            btn_find_devices.isEnabled = false
        } else if (bluetoothAdapter!!.isEnabled) {
            s_bluetooth.isChecked = true
            btn_find_devices.isEnabled = true
        }
    }

    private fun initReceiver() {
        val filter = IntentFilter()
        filter.addAction(BluetoothDevice.ACTION_FOUND)
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_STARTED)
        filter.addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
        activity?.registerReceiver(receiver, filter)
    }

    private fun enableBluetooth() {
        if (!bluetoothAdapter!!.isEnabled) {
            val intent = Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE)
            startActivityForResult(intent, ENABLE_REQUEST_CODE)
        }
    }

    private fun disableBluetooth() {
        if (bluetoothAdapter!!.isEnabled) {
            bluetoothAdapter?.disable()
            btn_find_devices.isEnabled = false
        }
    }

    private fun findDevices() {
        bluetoothAdapter?.startDiscovery()
        adapter?.clear()
    }
}